{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Deliverables:  36 total points\n",
    "\n",
    " - seqCleaner  9 points\n",
    " - fastqParse  9 points\n",
    " - coordinateMathSoln 9 points\n",
    " - converter 9 points\n",
    " - Optional: Triad class rewrite (5 points extra credit)\n",
    " \n",
    " These must be submitted as a combined jupyter file and as 4(or 5) .py files\n",
    "     \n",
    " Due: Monday April 16, 2018 11:55pm\n",
    "\n",
    "# Lab 2 – Manipulating Data Types (36 points)\n",
    "## Overview\n",
    "Last week we introduced the Python function input(), which is used to take in string data. This week we will use input() to take in string and numeric (float and integer) data. \n",
    "\n",
    "Remember that all data returned from input() is a string object, so this will mean that you need to convert any numeric data to their respective numeric objects. The exercises in this lab will give you practice manipulating various types of data that commonly arise in computational biology problems.\n",
    "\n",
    "We are again using jupyter for this assignment, and as a final step we will create distinct command line programs from each of the parts.\n",
    "Please save each of the four (or five) python programs as .py files into your LAB02 folder. You can copy the text from the code cell, then paste it into an editor like notepad or textedit. Save that new file with the appropriate name ( eg. seqCleaner.py) into your LAB02 folder. Using terminal or cmd, you can now do a final test on that new program with the command:<br>\n",
    "python3 seqCleaner <br>\n",
    "\n",
    "All together, you will submit:\n",
    " - the jupyter file (Lab02.ipynb), \n",
    " - seqCleaner.py, \n",
    " - fastqParse,py, \n",
    " - coordinateMathSoln.py, \n",
    " - converter.py, and \n",
    " - (optionally) Triad.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence cleanup\n",
    "In this exercise, you will create a program to “clean up” a sequence of DNA by removing ambiguous bases (denoted by “N”) output from a sequencer. Your task is to create a Python program called seqCleaner that\n",
    " - asks for and collects a sequence of DNA using input()\n",
    " - removes the ambiguous parts of the sequence, outputs the “cleaned” sequence, replacing the ambiguous parts with a count in {}’s.\n",
    "For example, if I enter the sequence of DNA “AaNNNNNNGTC” (without quotes), the program will output:<br>\n",
    "AA{6}GTC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hints: \n",
    " - The input sequence is not guaranteed to be uppercase, but should be interpreted as though it is all uppercase.\n",
    " - Only the letters (A,C,G,T,N) will be included in the input.\n",
    " - Only the DNA characters (A,C,G,T) should remain after the cleanup. \n",
    " - The input may include, at most, one block of 'N' characters. \n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and specific line or block comments that explain the semantics of your implementation.\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs. This should be in the form of a  program level dosctring."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### seqCleaner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter DNA data: AaNNNNNNGTC\n",
      "\n",
      "seqCleaner does NOT remove special characters or numbers from the input.\n",
      "\n",
      "Cleaned DNA data: AA{6}GTC\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Andrew Zarzar (azarzar) \n",
    "# Group Members: None\n",
    "\n",
    "'''\n",
    "Reads a DNA string from user input and return a collapsed substring of embedded Ns to: {count}\n",
    "Any lower case letters are converted to uppercase\n",
    "Input: a DNA sequence that needs cleaning. (special characters and numbers will not be removed from the final cleaning)\n",
    "Output: a cleaned DNA sequence that replaces all non-A,T,G,C characters with {numberof non-characters}\n",
    "*** program will clean any dna string, special characters and numbers will not be removed from the dna string\n",
    "'''\n",
    "class DNAstring (str):\n",
    "    '''returns the length of the string'''\n",
    "    def length (self):\n",
    "        return (len(self))\n",
    "    '''makes dna string all lowercase, only upercases A,T,C,G characters. calls purify method'''\n",
    "    def sift(self):\n",
    "        dna = self\n",
    "        dna = dna.lower()\n",
    "        dna = dna.replace(\"a\", \"A\")\n",
    "        dna = dna.replace(\"t\", \"T\")\n",
    "        dna = dna.replace(\"c\", \"C\")\n",
    "        dna = dna.replace(\"g\",\"G\")\n",
    "        newDNA = self.purify(dna) # purify method is called to clean up the dna string\n",
    "        return newDNA # returns cleaned DNA string\n",
    "    \n",
    "    '''creates a list, counts non- ATGC characters and cleans sequence\n",
    "        Purification removes all non-capitalized letters, replaces it with an updated count enclosed in {}. As the count is updated,\n",
    "        the old value that was replaced in the previous loop iteration is removed from the list. \n",
    "    '''\n",
    "    def purify(self, sif):\n",
    "        dna = sif\n",
    "        l = list(dna) # creats list of dna\n",
    "        count = 0\n",
    "        nCount = 1 # the number of non A,T,G,C characters, resets after every A,T,G,C value found\n",
    "        while count < len(sif): # starts loop that loops through whole input\n",
    "            c = l[count]\n",
    "            d = str(l[count-1])\n",
    "            if c.islower():# if list value at index count isnt A,T,G,C, statement proceeds\n",
    "                l[count] = \"{\"+ str(nCount) +\"}\" #value at index count is turned into \"{nCount}\"\n",
    "                if (count!=0) and (d == \"{\"+ str(nCount-1) +\"}\"): l[count-1] = \"\" # statement erases old \"{nCount}\" value from list\n",
    "                count = count+1\n",
    "                nCount = nCount+1\n",
    "            else:\n",
    "                nCount = 1 # reset after every A,T,G,C, value found.\n",
    "                count = count+1      \n",
    "        return \"\".join(l)\n",
    "                \n",
    "''' Get user DNA data and clean it up.'''      \n",
    "def main():     \n",
    "    data = input('Enter DNA data: ')\n",
    "    thisDNA = DNAstring (data)\n",
    "    siftDNA = thisDNA.sift() #sifts DNA and calls purify method. Returns the results of the purification\n",
    "    print (\"\")\n",
    "    print(\"seqCleaner does NOT remove special characters or numbers from the input.\\n\")\n",
    "    print (\"Cleaned DNA data: \" + siftDNA) # prints cleaned DNA\n",
    "main()\n",
    "\n",
    "\n",
    "\n",
    "#https://stackoverflow.com/questions/10953189/count-lower-case-characters-in-a-string\n",
    "    # - provided guidance for implemenation of lowercase character identifier\n",
    "    \n",
    "#https://stackoverflow.com/questions/41752946/replacing-a-character-from-a-certain-index\n",
    "    # - provided guidance for replacing items in a list\n",
    "    \n",
    "#https://stackoverflow.com/questions/19970532/how-to-check-a-string-for-a-special-character\n",
    "    # - provided guidance for implementation of special character identification. (NOT IMPLEMENTED IN CODE ABOVE)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence information parsing\n",
    "In this exercise, you will create a program to “parse” sequence name information from a single line of a FASTQ formatted file. Your task is to create a Python script called fastqParse that:\n",
    " - asks for and collects the seqname line of a FASTQ file using input()\n",
    " - parses out each field of the run information from the string and displays each of them on a new line\n",
    "For example, if I enter the FASTQ seqname line:<br> \n",
    "__*@EAS139:136:FC706VJ:2:2104:15343:197393*__\n",
    "then the program will output:<br>\n",
    "Instrument = EAS139<br>\n",
    "Run ID = 136<br>\n",
    "Flow Cell ID = FC706VJ<br>\n",
    "Flow Cell Lane = 2<br>\n",
    "Tile Number = 2104<br>\n",
    "X-coord = 15343<br>\n",
    "Y-coord = 197393\n",
    "### Hints: \n",
    " - The input string is guaranteed to have 7 sequence name fields. \n",
    " - The first character of the FASTQ seqname line is “@” and each field of the run information is separated by a colon“:”.  \n",
    " - A reasonable solution would be around 16 lines of code excluding comments.\n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain documentation/comments\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### fastqParse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter FASTQ data: @EAS139:136:FC706VJ:2:2104:15343:197393\n",
      "Instrument = @EAS139\n",
      "Run ID = 136\n",
      "Flow Cell ID = FC706VJ\n",
      "Flow Cell Lane = 2\n",
      "Tile Number = 2104\n",
      "X-coord = 15343\n",
      "Y-coord = 197393\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Andrew Zarzar (azarzar) \n",
    "# Group Members: None\n",
    "\n",
    "'''\n",
    "Program parses through the string looking for colons, then seperates info and prints categorized info.\n",
    "***I know there are easier ways to do the same thing using .split()\n",
    "Input: fastq string on one line, seperated by colons\n",
    "Output: fastq information, grouped with respective description/identifier\n",
    "\n",
    "Assumptions: Colons are used to seperate information. There are no beginning or ending colons.\n",
    "'''\n",
    "\n",
    "class FastqString (str):\n",
    "    ''' Class has 3 methods that are each used to parse through different parts of the string'''\n",
    "    count = 0\n",
    "    def parseFirst(self):\n",
    "        ''' parses through the beginning of the string, looking for seperating semicolons. instantiates and updates count. returns first segment'''\n",
    "        global count\n",
    "        count = 0\n",
    "        data = self\n",
    "        x = data.index(\":\", count)\n",
    "        y = count\n",
    "        count += x+1\n",
    "        return self[y:x]\n",
    "    \n",
    "    def parseRest(self):\n",
    "        ''' parses through the middle of the string, looking for seperating semicolons while updating count. returns once a segment is found'''\n",
    "        global count\n",
    "        data = self\n",
    "        x = data.index(\":\", count)\n",
    "        y = count\n",
    "        count = x+1\n",
    "        return self[y:x]\n",
    "    \n",
    "    def parseEnd(self):\n",
    "        ''' uses count to parse through the end of the string, as data.index wont work for the end of the string'''\n",
    "        global count\n",
    "        data = self  \n",
    "        x = len(data)\n",
    "        y = count\n",
    "        count = x+1\n",
    "        return self[y:x]\n",
    "    \n",
    "def main():\n",
    "    ''' Asks for input, prints the outpus as the input is being parsed for data segments'''\n",
    "    data = input('Enter FASTQ data: ')\n",
    "    string = FastqString (data)\n",
    "    # The methods are used to parse through the string, count is constantly being updated as the methods are used\n",
    "    #The proper info is being printed\n",
    "    print (\"Instrument = \" +string.parseFirst())\n",
    "    print (\"Run ID = \"+string.parseRest())\n",
    "    print (\"Flow Cell ID = \"+string.parseRest())\n",
    "    print (\"Flow Cell Lane = \"+string.parseRest())\n",
    "    print (\"Tile Number = \"+string.parseRest())\n",
    "    print (\"X-coord = \"+string.parseRest())\n",
    "    print (\"Y-coord = \"+string.parseEnd())\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein coordinates\n",
    "In this exercise, you will create a program that takes three sets of atomic coordinates, all provided on a **single line**. The program then calculates the bond lengths and angles. For this program, you can start with the Triad class (provided). Your task is to create a Python program called coordinateMathSoln that:\n",
    " - asks for and collects three sets of coordinates using input(), only use 1 line for this data !!\n",
    " - outputs the N-C and N-Ca bond lengths and the C-N-Ca bond angle with correct number of significant digits (see below)\n",
    "For example, if I enter the following coordinates (notice.. they are all on one line !!!) :<br>\n",
    "\n",
    "C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465) <br>\n",
    "\n",
    "then the program will output the following three lines:<br>\n",
    "N-C bond length = 1.33<br>\n",
    "N-Ca bond length = 1.46<br>\n",
    "C-N-Ca bond angle = 124.0<br> \n",
    "(Note: make sure that the angle returned is in **degrees** !!)\n",
    "### Hints: \n",
    " - Each coordinate will contain only 3 numeric values. \n",
    " - Bond lengths are the distance between two points, which for points P and Q in 3-space,\n",
    "$(P_x , P_y , P_z )$ and $(Q_x , Q_y , Q_z )$ respectively, the distance between them is:\n",
    "\n",
    "\n",
    "$$\n",
    "\\begin {eqnarray}\n",
    "     \\|PQ\\| &=&\\sqrt{(P_x-Q_x)^2 + (P_y-Q_y)^2 + (P_z-Q_z)^2)} \\\\\n",
    "            &=&\\sqrt{\\sum^{i \\in x,y,z}{(P_i-Q_i)^2}} \\\\\n",
    "\\end{eqnarray}\n",
    "$$\n",
    "\n",
    " - Bond angles can be calculated from the dot product.\n",
    " \n",
    "Let’s say that we have three points in space labeled P, Q and R. We are interested in the angle at point Q that is made from line segments QP and QR.\n",
    "The dot product tells us (for standard vectors P and R) that:\n",
    "$$P \\cdot R=\\|P\\| \\|R\\| \\cos{\\theta}$$\n",
    "in this notation,$\\|P\\|$ refers to the length of vector P as a standard vector (assumed to begin at the origin (0,0,0) ).\n",
    "We can then see that the angle between vectors P and R can be found by:\n",
    "$$\\cos{\\theta}=\\frac{P \\cdot R} {\\|P\\| \\|R\\|}$$\n",
    "\n",
    "We can calculate the dot product using the sum of products of the vector components:\n",
    "$$P \\cdot R = \\sum^{i \\in x,y,z}{P_i R_i} $$\n",
    "\n",
    "Now, to find vector P in standard form, we need to remember that QP starts at Q, so we need to place the origin at Q and find out where P is in that new space. We do that by subtracting the components of Q from P.\n",
    "Putting all of this together, we get:\n",
    "$$\\theta=\\cos^{-1}{\\frac{\\sum^{i \\in x,y,z}{(P_i - Q_i) (R_i - Q_i) }} {\\|QP\\| \\|QR\\|}}$$\n",
    "Remember, $\\theta$ is in radians.\n",
    "\n",
    "Below I have given you a class (Triad) with methods to calculate dot products (dot), dot products of translated vectors (ndot), distances (dPQ, dPR, dPR) and angles in radians (angleP, angleQ and angleR) for each of the three points in a Triad object.\n",
    "A reasonable solution for this exercise involves around 12 lines of additional code excluding comments.\n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and line comments (using #)\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs as a program level docstring\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### coordinateMathSoln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter 3 atomic coordinates on the same line: \n",
      "C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465)\n",
      "\n",
      "N-C bond length = 1.33\n",
      "N-Ca bond length = 1.46\n",
      "C-N-Ca bond angle = 124.0\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Andrew Zarzar (azarzar) \n",
    "# Group Members: None\n",
    "\n",
    "'''\n",
    "Program uses the Triad class to perform calculations that returns the bond length between point1 and point2, and \n",
    "the bond length between point 2 and point3. The bond angle between point1-point2-point3 is then returned. Triad methods are used\n",
    "for all calculations and main() is used for proper formating of inputs. \n",
    "\n",
    "Assumptions: There are only 3 points.\n",
    "'''\n",
    "\n",
    "import math\n",
    "class Triad :\n",
    "    \"\"\"\n",
    "    Calculate angles and distances among a triad of points.\n",
    " \n",
    "    Author: David Bernick\n",
    "    Date: March 21, 2013\n",
    "    Points can be supplied in any dimensional space as long as they are consistent.\n",
    "    Points are supplied as tupels in n-dimensions, and there should be three\n",
    "    of those to make the triad. Each point is positionally named as p,q,r\n",
    "    and the corresponding angles are then angleP, angleQ and angleR.\n",
    "    Distances are given by dPQ(), dPR() and dQR()\n",
    " \n",
    "    Required Modules: math\n",
    "    initialized: 3 positional tuples representing Points in n-space\n",
    "             p1 = Triad( p=(1,0,0), q=(0,0,0), r=(0,1,0) )\n",
    "    attributes: p,q,r the 3 tuples representing points in N-space\n",
    "    methods:  angleP(), angleR(), angleQ() angles measured in radians\n",
    "          dPQ(), dPR(), dQR() distances in the same units of p,q,r\n",
    " \n",
    "    \"\"\"\n",
    " \n",
    "    def __init__(self,p,q,r) :\n",
    "        \"\"\" Construct a Triad. \n",
    "        \n",
    "        Example construction:\n",
    "            p1 = Triad( p=(1.,0.,0.), q=(0.,0.,0.), r=(0.,0.,0.) ). \n",
    "        \"\"\"\n",
    "        self.p = p\n",
    "        self.q = q\n",
    "        self.r = r\n",
    "# private helper methods\n",
    "    def d2 (self,a,b) : # calculate squared distance of point a to b\n",
    "        return float(sum((ia-ib)*(ia-ib)  for  ia,ib in zip (a,b)))\n",
    "    \n",
    "    def dot (self,a,b) : # dotProd of standard vectors a,b\n",
    "        return float(sum(ia*ib for ia,ib in zip(a,b)))\n",
    "    \n",
    "    def ndot (self,a,b,c) : # dotProd of vec. a,c standardized to b\n",
    "        return float(sum((ia-ib)*(ic-ib) for ia,ib,ic in zip (a,b,c)))\n",
    "    \n",
    "# calculate lengths(distances) of segments PQ, PR and QR\n",
    "    def dPQ (self):\n",
    "        \"\"\" Provides the distance between point p and point q \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.q))\n",
    "    \n",
    "    def dPR (self):\n",
    "        \"\"\" Provides the distance between point p and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.r))\n",
    "    \n",
    "    def dQR (self):\n",
    "        \"\"\" Provides the distance between point q and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.q,self.r))\n",
    "    \n",
    "    def angleP (self) :\n",
    "        \"\"\" Provides the angle made at point p by segments pq and pr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.q,self.p,self.r) /   math.sqrt(self.d2(self.q,self.p)*self.d2(self.r,self.p)))\n",
    "    \n",
    "    def angleQ (self) :\n",
    "        \"\"\" Provides the angle made at point q by segments qp and qr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.q,self.r) /  math.sqrt(self.d2(self.p,self.q)*self.d2(self.r,self.q)))\n",
    " \n",
    "    def angleR (self) :\n",
    "        \"\"\" Provides the angle made at point r by segments rp and rq (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.r,self.q) /  math.sqrt(self.d2(self.p,self.r)*self.d2(self.q,self.r)))\n",
    "\n",
    "\"\"\"Houses the semantics of the program. Sorts through input, and performs calculations using Triad class\"\"\"\n",
    "def main():\n",
    "    data = input(\"Please enter 3 atomic coordinates on the same line: \\n\")\n",
    "    data = data.split(\"=\") #splits input at \"=\" to start the seperation of points\n",
    "    data = \"\".join(data) # data is joined back into a string after the split\n",
    "    data = data.replace(\" \",\"\") #data removes all white space from the string.\n",
    "    \n",
    "    #indexes the location of the first point data using the locations of parenthesis\n",
    "    p1start = data.index(\"(\")\n",
    "    p1end = data.index(\")\")\n",
    "    #indexes the location of the second point data using the locations of parenthesis and first point data\n",
    "    p2start = data.index(\"(\", p1end)\n",
    "    p2end = data.index(\")\", p2start)\n",
    "    #indexes the location of the third point data using the locations of parenthesis and second point data\n",
    "    p3start = data.index(\"(\", p2end)\n",
    "    p3end = data.index(\")\", p3start)\n",
    "    #The names of the elements corresponding to the points are found and stored using the relative location of the parenthesis.\n",
    "    Element1 = data[0:p1start]\n",
    "    Element2 = data[p1end+1:p2start]\n",
    "    Element3 = data[p2end+1:p3start]\n",
    "    #stores the values of coordinates without including parenthesis\n",
    "    coord1 = data[p1start+1:p1end]\n",
    "    coord2 = data[p2start+1:p2end]\n",
    "    coord3 = data[p3start+1:p3end]\n",
    "    #splits string of each coordinate at commas, to make a list for each of the coordinates\n",
    "    Lcoord1 = coord1.split(',')\n",
    "    Lcoord2 = coord2.split(',')\n",
    "    Lcoord3 = coord3.split(',')\n",
    "    # casts the type of of every value in the 3 lists from a string into a float\n",
    "    for i in range(len(Lcoord1)):\n",
    "        Lcoord1[i] = float(Lcoord1[i])\n",
    "    for i in range(len(Lcoord2)):\n",
    "        Lcoord2[i] = float(Lcoord2[i])\n",
    "    for i in range(len(Lcoord1)):\n",
    "        Lcoord3[i] = float(Lcoord3[i])\n",
    "        \n",
    "    \"\"\"#debugging test\n",
    "    print (data)\n",
    "    print (Element1)\n",
    "    print (Element2)\n",
    "    print (Element3)\n",
    "    print(coord1)\n",
    "    print(coord2)\n",
    "    print(coord3)\n",
    "    print(Lcoord1)\n",
    "    print(Lcoord2)\n",
    "    print(Lcoord3)\n",
    "    \"\"\"\n",
    "    #the float values are put into the triad class\n",
    "    coords = Triad(Lcoord1,Lcoord2,Lcoord3)\n",
    "    # triad methods are used to calculated bond lengths between PQ and QR. Bond angle is also calcuated\n",
    "    bondLengthqp = coords.dPQ()\n",
    "    bondLengthqr = coords.dQR()\n",
    "    bondAngle = math.degrees(coords.angleQ())\n",
    "    print(\"\")\n",
    "    #Results are printed using correct formating\n",
    "    print (Element2 +\"-\"+Element1 + \" bond length = \"+ \"{0:0.2f}\".format(bondLengthqp))\n",
    "    print (Element2 +\"-\"+Element3 + \" bond length = \"+ \"{0:0.2f}\".format(bondLengthqr))\n",
    "    print (Element1 +\"-\"+Element2+\"-\"+Element3 + \" bond angle = \"+ \"{0:0.1f}\".format(bondAngle))\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extra credit (5 points): Rewrite the Triad class.  \n",
    "For extra-credit, provide a direct replacement for the Triad class. The external methods that calculate angles, distances, and points (tuples) p,q and r must be maintained such that either version of the Triad class can be used.\n",
    "\n",
    "You could use the cosine law to calculate angles instead of the dot product.  You might make use of the numpy module. You might recode each of the methods to avoid using zip. You might consider using list iterations.\n",
    "Your Triad replacement must reimplement all of Triad public function, without using zip and without being a trivial rewrite. Your implementation need not be as compact as the current implementation, and it needs to be correct and fully documented to receive full credit. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Triad rewrite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Triad:\n",
    "    def __init__(self,p,q,r) :\n",
    "        \"\"\" Construct a Triad.  \n",
    "        \n",
    "        Example object construction:\n",
    "            p1 = Triad( p=(1.,0.,0.), q=(0.,0.,0.), r=(0.,0.,0.) ). \n",
    "        \"\"\"\n",
    "        self.p = p\n",
    "        self.q = q\n",
    "        self.r = r\n",
    "    def dPQ (self):\n",
    "        \"\"\" Provides the distance between point p and point q \"\"\"\n",
    "        pass\n",
    "    \n",
    "    def dPR (self):\n",
    "        \"\"\" Provides the distance between point p and point r \"\"\"\n",
    "        pass\n",
    "    \n",
    "    def dQR (self):\n",
    "        \"\"\" Provides the distance between point q and point r \"\"\"\n",
    "        pass\n",
    "    \n",
    "    def angleP (self) :\n",
    "        \"\"\" Provides the angle made at point p by segments pq and pr (radians). \"\"\"\n",
    "        pass\n",
    "    \n",
    "    def angleQ (self) :\n",
    "        \"\"\" Provides the angle made at point q by segments qp and qr (radians). \"\"\"\n",
    "        pass\n",
    "        \n",
    "    def angleR (self) :\n",
    "        \"\"\" Provides the angle made at point r by segments rp and rq (radians). \"\"\"\n",
    "        pass\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Codon tables and amino acid letter converters\n",
    "In this exercise, you will create a program that uses mappings to convert sequence information between different amino acid representations.  This includes the 3-letter codon code (RNA and DNA), the one letter amino acid code and the 3-letter amino acid code.\n",
    "The program will use different dictionaries that represent :\n",
    "codon tables – one for DNA and the other for RNA, and\n",
    "amino acid letter representation converters. \n",
    "\n",
    "Your task is to create a Python program called converter that\n",
    "asks for and collects a single input string using input()\n",
    "parses the string, looks up the information in the appropriate dictionary, and outputs the correct conversion\n",
    "For example:<br>\n",
    "\n",
    "if I enter “ATG” (without quotes), then the program will output:<br>\n",
    "ATG = MET<br>\n",
    " \n",
    "if I enter “UAG” (without quotes), then the program will output:<br>\n",
    "UAG = ---<br>\n",
    "if I enter “E” (without quotes), then the program will output:<br>\n",
    "E = GLU<br>\n",
    "if I enter “Asp” (without quotes), then the program will output:<br>\n",
    "ASP = D<br>\n",
    "### Hints: \n",
    " - The program might not get a valid codon.  In that case, it should output 'unknown'.  You can use the dictionary method 'get' and include a default_value to handle the 'unknown' case.  See Model p. 72 for an example.\n",
    " \n",
    "To get full credit on this assignment, your program needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and line comments (using #)\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs as a program level docstring\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### converter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Instructions: Converter will convert DNA codons to amino acids, RNA codons to amino acids, and amino acid code to a different amino acid code format  \n",
      "Program WILL NOT yield results for a string containing incomplete codons\n",
      "1. Any string of DNA codons will be converted into an amino acid sequence  \n",
      "2. Any string or RNA codons will be converted into an amino acid sequence \n",
      "3. Any single amino acid will be converted to its alternate format \n",
      "Inputs are NOT case sensitive, and codons can be seperated by spaces,commas, or nothing.\n",
      "'UNKNOWN' or '???' will be outputted if codon is invalid\n",
      "\n",
      "Enter string: atg\n",
      "\n",
      "ATG = MET\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Andrew Zarzar (azarzar) \n",
    "# Group Members: None\n",
    "\n",
    "'''\n",
    "Program will give user instuctions and ask for input. \n",
    "    If DNA codons are inputed, the corresponding amino acids are outputed\n",
    "    If RNA codons are inputed, the corresponding amino acids are outputed\n",
    "    If a 1 letter amino acid code is inputed, the corresponding 3 letter amino acid code is outputed. Vice-versa\n",
    "    Program is not case sensitive, and can accept unseperated codons, or codons seperated by spaces or commas. \n",
    "    Codons cannot switch between cases\n",
    "    \n",
    "    Program differentiates the inputs using a series of specific if-else statements.\n",
    "    \n",
    "    *** I took the extra time to design a program that is more robust, and has the ability to translate whole sequences. \n",
    "'''\n",
    "# 3 letter Amino acid to 1 letter Amino acid dictionary\n",
    "short_AA = {\n",
    "            'CYS': 'C', 'ASP': 'D', 'SER': 'S', 'GLN': 'Q', 'LYS': 'K',\n",
    "            'ILE': 'I', 'PRO': 'P', 'THR': 'T', 'PHE': 'F', 'ASN': 'N', \n",
    "            'GLY': 'G', 'HIS': 'H', 'LEU': 'L', 'ARG': 'R', 'TRP': 'W', \n",
    "            'ALA': 'A', 'VAL': 'V', 'GLU': 'E', 'TYR': 'Y', 'MET': 'M'\n",
    "            }\n",
    "#creates new dictionary based on short_AA, but swaps items out to allow conversion from 1 AA to 3 AA code.\n",
    "long_AA = {value:key for key,value in short_AA.items()}\n",
    "\n",
    "#RNA codon to amino acid dictionary\n",
    "RNA_codon_table = {\n",
    "# Second Base\n",
    "# U             C             A             G\n",
    "#U\n",
    "'UUU': 'Phe', 'UCU': 'Ser', 'UAU': 'Tyr', 'UGU': 'Cys',\n",
    "'UUC': 'Phe', 'UCC': 'Ser', 'UAC': 'Tyr', 'UGC': 'Cys',\n",
    "'UUA': 'Leu', 'UCA': 'Ser', 'UAA': '---', 'UGA': '---',\n",
    "'UUG': 'Leu', 'UCG': 'Ser', 'UAG': '---', 'UGG': 'Trp',\n",
    "#C \n",
    "'CUU': 'Leu', 'CCU': 'Pro', 'CAU': 'His', 'CGU': 'Arg',\n",
    "'CUC': 'Leu', 'CCC': 'Pro', 'CAC': 'His', 'CGC': 'Arg',\n",
    "'CUA': 'Leu', 'CCA': 'Pro', 'CAA': 'Gln', 'CGA': 'Arg',\n",
    "'CUG': 'Leu', 'CCG': 'Pro', 'CAG': 'Gln', 'CGG': 'Arg',\n",
    "#A\n",
    "'AUU': 'Ile', 'ACU': 'Thr', 'AAU': 'Asn', 'AGU': 'Ser',\n",
    "'AUC': 'Ile', 'ACC': 'Thr', 'AAC': 'Asn', 'AGC': 'Ser',\n",
    "'AUA': 'Ile', 'ACA': 'Thr', 'AAA': 'Lys', 'AGA': 'Arg',\n",
    "'AUG': 'Met', 'ACG': 'Thr', 'AAG': 'Lys', 'AGG': 'Arg',\n",
    "#G\n",
    "'GUU': 'Val', 'GCU': 'Ala', 'GAU': 'Asp', 'GGU': 'Gly',\n",
    "'GUC': 'Val', 'GCC': 'Ala', 'GAC': 'Asp', 'GGC': 'Gly',\n",
    "'GUA': 'Val', 'GCA': 'Ala', 'GAA': 'Glu', 'GGA': 'Gly',\n",
    "'GUG': 'Val', 'GCG': 'Ala', 'GAG': 'Glu', 'GGG': 'Gly'\n",
    "}\n",
    "# creates DNA codon to amino acid dictionary based on RNA codon table\n",
    "dnaCodonTable = {key.replace('U','T'):value for key, value in RNA_codon_table.items()}\n",
    "\n",
    "'''\n",
    "method is used to loop through multiple rna codons and return amino acids. \n",
    "Creates list of every codon from string input.\n",
    "The list is looped through and replaced with amino acids from dictionary.\n",
    "The list is joined back into a string\n",
    "'''\n",
    "def rnaLoop(str):\n",
    "    codonList = [ str[i:i+3] for i in range(0, len(str), 3) ]#list of every 3 codons\n",
    "    translatedList = [RNA_codon_table.get(codonList[i].upper(),\"???\")+\"-\" for i in range(0, len(codonList))]#codon--> dictionary\n",
    "    x = \"\".join(translatedList)#joins list\n",
    "    return x[0:len(x)-1]#removes end dash\n",
    "\n",
    "'''\n",
    "method is used to loop through multiple dna codons and return amino acids. \n",
    "Creates list of every codon from string input.\n",
    "The list is looped through and replaced with amino acids from dictionary.\n",
    "The list is joined back into a string\n",
    "'''\n",
    "def dnaLoop(str):\n",
    "    codonList = [ str[i:i+3] for i in range(0, len(str), 3) ]#List of every 3 codons\n",
    "    translatedList = [dnaCodonTable.get(codonList[i].upper(),\"???\")+\"-\" for i in range(0, len(codonList))]#inputs codon into dictionary\n",
    "    x = \"\".join(translatedList)#joins list\n",
    "    return x[0:len(x)-1]#removes end dash\n",
    "\n",
    "def main():\n",
    "    print (\"Instructions: Converter will convert DNA codons to amino acids, RNA codons to amino acids, and amino acid code to a different amino acid code format  \\n\" + \n",
    "           \"Program WILL NOT yield results for a string containing incomplete codons\\n\"+\n",
    "           \"1. Any string of DNA codons will be converted into an amino acid sequence  \\n\"+ \n",
    "           \"2. Any string or RNA codons will be converted into an amino acid sequence \\n\"+\n",
    "           \"3. Any single amino acid will be converted to its alternate format \\n\" + \"Inputs are NOT case sensitive, and codons can be seperated by spaces,commas, or nothing.\\n\"+\n",
    "           \"'UNKNOWN' or '???' will be outputted if codon is invalid\\n\"   )\n",
    "    data = input(\"Enter string: \")\n",
    "    data = data.replace(\" \", \"\") #removes blank spaces from input\n",
    "    data = data.replace(\",\",\"\") #removes commas from input\n",
    "    print(\"\")\n",
    "    # if input is 3 characters long and is not RNA, then program will search dnaCodon and short_AA dictionaries\n",
    "    if len(data) ==3 and (data.isupper() or data.islower()) and (('T' in data) or ('t' in data)): print (data.upper()+ \" = \" + dnaCodonTable.get(data.upper(),short_AA.get(data.upper(),\"Unknown\")).upper())\n",
    "    # if input is 3 characters long, then program will search rnaCodon and short_AA dictionaries\n",
    "    elif len(data)==3 and (data.isupper()or data.islower()): print (data.upper()+ \" = \" + RNA_codon_table.get(data.upper(),short_AA.get(data.upper(),\"Unknown\")).upper())\n",
    "    #satement may be redundant, but will be kept just in case. \n",
    "    elif len(data)==3 and not data.isupper() and not data.islower(): print (data.upper()+ \" = \" + short_AA.get(data.upper(),\"Unknown\").upper())\n",
    "    # if input is 1 character long, then program will search long_AA dictionary to get 3 letter amino acids  \n",
    "    elif len(data)==1: print (data.upper()+ \" = \" + long_AA.get(data.upper(),\"Unknown\").upper())\n",
    "    # if input is a complete set of codons, and is not RNA, then the codons will be converted to 3 letter amino acids\n",
    "    elif len(data)% 3==0 and (('T' in data) or ('t' in data)): return dnaLoop(data)\n",
    "    # if input is a complete set of codons, then the codons will be converted to 3 letter amino acids\n",
    "    elif len(data)% 3==0: return rnaLoop(data)\n",
    "    # if input isn't a complete set of codons, or a single letter amino acids, then unknown will be returned\n",
    "    else: print(data+\" = UNKNOWN\") \n",
    "# If dictionary searches turn up negative, then an unknown or ??? value will be returned\n",
    "main()\n",
    "\n",
    "# Method rnaLoop based on https://stackoverflow.com/questions/5711452/how-do-i-slice-a-string-every-3-indices\n",
    "# Method dnaLoop based on rnaLoop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit your code and answers\n",
    " \n",
    "Important: please save your work before logging out of the computer. This will ensure that you have a copy of your work and you will avoid having to redo everything in the  event that something happens with the lab computers. The two recommended solutions (pick one) at this point are to:\n",
    " - Email your code files to yourself\n",
    " - Copy your code files to a pen drive\n",
    "\n",
    "The class will be using canvas to submit the source code files created as part of this lab assignment.\n",
    "For this lab, you should upload the following programs as attachments:\n",
    " - the jupyter file (Lab02.ipynb), \n",
    "\n",
    " - seqCleaner.py 9 points\n",
    " - fastqParse.py 9 points\n",
    " - coordinateMathSoln.py 9 points\n",
    " - converter.py 9 points\n",
    " - triad.py Triad rewrite (optional) 5 points\n",
    " \n",
    "Important: to get full credit on this lab assignment, each of the code files you submit needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain proper docstrings and appropriate line style comments (#)\n",
    " \n",
    "Congratulations, you finished your second lab assignment!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
